### ISSUE TITLE
Add implicit parameters to language

### ISSUE THREAD

**Issue by MatthewDaggitt** [vehicle-lang/vehicle#92]
Parameters that should be inferred from the loaded resources, e.g. neural network and datasets. Needed for expressing classifcation robustness.

Suggested syntax: 
```
parameter {n : Nat}
implicit parameter n : Nat
```

Implementation: pretty trivial for datasets, just desugars to a hole that should be inferred by the type-checker automatically once the resources have been expanded. More difficult when used in networks.

**Comment by MatthewDaggitt** [comment]
Yet another issue blocked by normalisation in the type-checker.
